File: fontes/backend/php/fenix/src/model/cadastros/Pessoa.php
<?php
/*******************************************************************************
Title: T2Ti ERP 3.0
Description: Model relacionado à tabela [PESSOA] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
declare(strict_types=1);

class Pessoa extends EloquentModel implements \JsonSerializable
{
    /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table = 'PESSOA';

    /**
     * Eager Loading - Relacionamentos que devem ser sempre carregados por padrão
     *
     * @var array
     */
    protected $with = ['cliente', 'colaborador', 'listaPessoaContato', 'pessoaFisica', 'pessoaJuridica', ];
	
    /**
     * Relations
     */
    public function cliente()
    {
    	return $this->hasOne(Cliente::class, 'ID_PESSOA', 'ID');
    }

    public function colaborador()
    {
    	return $this->hasOne(Colaborador::class, 'ID_PESSOA', 'ID');
    }

    public function listaPessoaContato()
    {
    	return $this->hasMany(PessoaContato::class, 'ID_PESSOA', 'ID');
    }

    public function pessoaFisica()
    {
    	return $this->hasOne(PessoaFisica::class, 'ID_PESSOA', 'ID');
    }

    public function pessoaJuridica()
    {
    	return $this->hasOne(PessoaJuridica::class, 'ID_PESSOA', 'ID');
    }


    /**
     * Gets e Sets
     */
    public function getIdAttribute() 
	{
		return $this->attributes['ID'];
	}

	public function setIdAttribute($id) 
	{
		$this->attributes['ID'] = $id;
	}

    public function getNomeAttribute() 
	{
		return $this->attributes['NOME'];
	}

	public function setNomeAttribute($nome) 
	{
		$this->attributes['NOME'] = $nome;
	}

    public function getTipoAttribute() 
	{
		return $this->attributes['TIPO'];
	}

	public function setTipoAttribute($tipo) 
	{
		$this->attributes['TIPO'] = $tipo;
	}

    public function getSiteAttribute() 
	{
		return $this->attributes['SITE'];
	}

	public function setSiteAttribute($site) 
	{
		$this->attributes['SITE'] = $site;
	}

    public function getEmailAttribute() 
	{
		return $this->attributes['EMAIL'];
	}

	public function setEmailAttribute($email) 
	{
		$this->attributes['EMAIL'] = $email;
	}

    public function getEhClienteAttribute() 
	{
		return $this->attributes['EH_CLIENTE'];
	}

	public function setEhClienteAttribute($ehCliente) 
	{
		$this->attributes['EH_CLIENTE'] = $ehCliente;
	}

    public function getEhFornecedorAttribute() 
	{
		return $this->attributes['EH_FORNECEDOR'];
	}

	public function setEhFornecedorAttribute($ehFornecedor) 
	{
		$this->attributes['EH_FORNECEDOR'] = $ehFornecedor;
	}

    public function getEhTransportadoraAttribute() 
	{
		return $this->attributes['EH_TRANSPORTADORA'];
	}

	public function setEhTransportadoraAttribute($ehTransportadora) 
	{
		$this->attributes['EH_TRANSPORTADORA'] = $ehTransportadora;
	}

    public function getEhColaboradorAttribute() 
	{
		return $this->attributes['EH_COLABORADOR'];
	}

	public function setEhColaboradorAttribute($ehColaborador) 
	{
		$this->attributes['EH_COLABORADOR'] = $ehColaborador;
	}

    public function getEhContadorAttribute() 
	{
		return $this->attributes['EH_CONTADOR'];
	}

	public function setEhContadorAttribute($ehContador) 
	{
		$this->attributes['EH_CONTADOR'] = $ehContador;
	}


    /**
     * Mapping
     */
    public function mapear($objeto)
    {
        if (isset($objeto)) {
            isset($objeto->id) ? $this->setIdAttribute($objeto->id) : $this->setIdAttribute(null);

			$this->setNomeAttribute($objeto->nome);
			$this->setTipoAttribute($objeto->tipo);
			$this->setSiteAttribute($objeto->site);
			$this->setEmailAttribute($objeto->email);
			$this->setEhClienteAttribute($objeto->ehCliente);
			$this->setEhFornecedorAttribute($objeto->ehFornecedor);
			$this->setEhTransportadoraAttribute($objeto->ehTransportadora);
			$this->setEhColaboradorAttribute($objeto->ehColaborador);
			$this->setEhContadorAttribute($objeto->ehContador);

			// vincular objetos
        }
    }

    /**
     * Serialization
     * {@inheritdoc}
     */
    public function jsonSerialize()
    {
        return [
			'id' => $this->getIdAttribute(),
			'nome' => $this->getNomeAttribute(),
			'tipo' => $this->getTipoAttribute(),
			'site' => $this->getSiteAttribute(),
			'email' => $this->getEmailAttribute(),
			'ehCliente' => $this->getEhClienteAttribute(),
			'ehFornecedor' => $this->getEhFornecedorAttribute(),
			'ehTransportadora' => $this->getEhTransportadoraAttribute(),
			'ehColaborador' => $this->getEhColaboradorAttribute(),
			'ehContador' => $this->getEhContadorAttribute(),
			'cliente' => $this->cliente,
			'colaborador' => $this->colaborador,
			'pessoaFisica' => $this->pessoaFisica,
			'pessoaJuridica' => $this->pessoaJuridica,
			'listaPessoaContato' => $this->listaPessoaContato,
        ];
    }
}

File: fontes/backend/delphi/dmvc/Model/Cadastros/Pessoa.pas
{*******************************************************************************
Title: T2Ti ERP Fenix
Description: Model relacionado à tabela [PESSOA]

The MIT License

Copyright: Copyright (C) 2020 T2Ti.COM

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation
files (the "Software"), to deal in the Software without
restriction, including without limitation the rights to use,
copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following
conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

       The author may be contacted at:
           t2ti.com@gmail.com

@author Albert Eije (alberteije@gmail.com)
@version 1.0.0
*******************************************************************************}
unit Pessoa;

interface

uses
  Generics.Collections, System.SysUtils,
  Cliente, Colaborador, Contador, Fornecedor, PessoaContato, PessoaEndereco, PessoaFisica, PessoaJuridica, PessoaTelefone, Transportadora,
  MVCFramework.Serializer.Commons, ModelBase;

type

  [MVCNameCase(ncCamelCase)]
  TPessoa = class(TModelBase)
  private
    FId: Integer;
    FNome: string;
    FTipo: string;
    FSite: string;
    FEmail: string;
    FEhCliente: string;
    FEhFornecedor: string;
    FEhTransportadora: string;
    FEhColaborador: string;
    FEhContador: string;

    FCliente: TCliente;
    FColaborador: TColaborador;
    FContador: TContador;
    FFornecedor: TFornecedor;
    FPessoaFisica: TPessoaFisica;
    FPessoaJuridica: TPessoaJuridica;
    FTransportadora: TTransportadora;
    FListaPessoaContato: TObjectList<TPessoaContato>;
    FListaPessoaEndereco: TObjectList<TPessoaEndereco>;
    FListaPessoaTelefone: TObjectList<TPessoaTelefone>;
  public
    constructor Create; virtual;
    destructor Destroy; override;

    [MVCColumnAttribute('ID', True)]
		[MVCNameAsAttribute('id')]
		property Id: Integer read FId write FId;
    [MVCColumnAttribute('NOME')]
		[MVCNameAsAttribute('nome')]
		property Nome: string read FNome write FNome;
    [MVCColumnAttribute('TIPO')]
		[MVCNameAsAttribute('tipo')]
		property Tipo: string read FTipo write FTipo;
    [MVCColumnAttribute('SITE')]
		[MVCNameAsAttribute('site')]
		property Site: string read FSite write FSite;
    [MVCColumnAttribute('EMAIL')]
		[MVCNameAsAttribute('email')]
		property Email: string read FEmail write FEmail;
    [MVCColumnAttribute('EH_CLIENTE')]
		[MVCNameAsAttribute('ehCliente')]
		property EhCliente: string read FEhCliente write FEhCliente;
    [MVCColumnAttribute('EH_FORNECEDOR')]
		[MVCNameAsAttribute('ehFornecedor')]
		property EhFornecedor: string read FEhFornecedor write FEhFornecedor;
    [MVCColumnAttribute('EH_TRANSPORTADORA')]
		[MVCNameAsAttribute('ehTransportadora')]
		property EhTransportadora: string read FEhTransportadora write FEhTransportadora;
    [MVCColumnAttribute('EH_COLABORADOR')]
		[MVCNameAsAttribute('ehColaborador')]
		property EhColaborador: string read FEhColaborador write FEhColaborador;
    [MVCColumnAttribute('EH_CONTADOR')]
		[MVCNameAsAttribute('ehContador')]
		property EhContador: string read FEhContador write FEhContador;

    [MVCNameAsAttribute('cliente')]
		property Cliente: TCliente read FCliente write FCliente;
    [MVCNameAsAttribute('colaborador')]
		property Colaborador: TColaborador read FColaborador write FColaborador;
    [MVCNameAsAttribute('contador')]
		property Contador: TContador read FContador write FContador;
    [MVCNameAsAttribute('fornecedor')]
		property Fornecedor: TFornecedor read FFornecedor write FFornecedor;
    [MVCNameAsAttribute('pessoaFisica')]
		property PessoaFisica: TPessoaFisica read FPessoaFisica write FPessoaFisica;
    [MVCNameAsAttribute('pessoaJuridica')]
		property PessoaJuridica: TPessoaJuridica read FPessoaJuridica write FPessoaJuridica;
    [MVCNameAsAttribute('transportadora')]
		property Transportadora: TTransportadora read FTransportadora write FTransportadora;
    [MapperListOf(TPessoaContato)]
		[MVCNameAsAttribute('listaPessoaContato')]
		property ListaPessoaContato: TObjectList<TPessoaContato> read FListaPessoaContato write FListaPessoaContato;
    [MapperListOf(TPessoaEndereco)]
		[MVCNameAsAttribute('listaPessoaEndereco')]
		property ListaPessoaEndereco: TObjectList<TPessoaEndereco> read FListaPessoaEndereco write FListaPessoaEndereco;
    [MapperListOf(TPessoaTelefone)]
		[MVCNameAsAttribute('listaPessoaTelefone')]
		property ListaPessoaTelefone: TObjectList<TPessoaTelefone> read FListaPessoaTelefone write FListaPessoaTelefone;
  end;

implementation

{ TPessoa }

constructor TPessoa.Create;
begin
  inherited;
  FCliente := TCliente.Create;
  FColaborador := TColaborador.Create;
  FContador := TContador.Create;
  FFornecedor := TFornecedor.Create;
  FListaPessoaContato := TObjectList<TPessoaContato>.Create;
  FListaPessoaEndereco := TObjectList<TPessoaEndereco>.Create;
  FPessoaFisica := TPessoaFisica.Create;
  FPessoaJuridica := TPessoaJuridica.Create;
  FListaPessoaTelefone := TObjectList<TPessoaTelefone>.Create;
  FTransportadora := TTransportadora.Create;
end;

destructor TPessoa.Destroy;
begin
  FreeAndNil(FCliente);
  FreeAndNil(FColaborador);
  FreeAndNil(FContador);
  FreeAndNil(FFornecedor);
  FreeAndNil(FListaPessoaContato);
  FreeAndNil(FListaPessoaEndereco);
  FreeAndNil(FPessoaFisica);
  FreeAndNil(FPessoaJuridica);
  FreeAndNil(FListaPessoaTelefone);
  FreeAndNil(FTransportadora);
  inherited;
end;

end.
